File: src/server/processManager.ts
import type { Subprocess } from 'bun';
import { validateDevCommand } from './security.js';

interface ManagedProcess {
  proc: Subprocess;
  projectId: string;
  projectPath: string;
  command: string;
  buffer: string[];
  listeners: Set<ReadableStreamDefaultController>;
  startedAt: number;
  autoRestart: boolean;
  restartCount: number;
  crashTimes: number[];
  restarting: boolean;
}

export interface ProcessInfo {
  projectId: string;
  running: boolean;
  pid: number | null;
  startedAt: number;
  autoRestart: boolean;
  restartCount: number;
}

const MAX_BUFFER = 500;
const MAX_CONCURRENT = 20;
const MAX_RESTARTS = 3;
const RESTART_WINDOW_MS = 5 * 60 * 1000; // 5 minutes
const processes = new Map<string, ManagedProcess>();

export function startProcess(
  projectId: string,
  path: string,
  command: string,
  options?: { autoRestart?: boolean }
): boolean {
  // Validate command against allowlist
  const cmdCheck = validateDevCommand(command);
  if (!cmdCheck.valid) {
    console.warn(`[processManager] Blocked unsafe command: ${command}`);
    return false;
  }

  // Limit concurrent processes
  if (processes.size >= MAX_CONCURRENT) {
    console.warn(`[processManager] Max concurrent processes (${MAX_CONCURRENT}) reached`);
    return false;
  }

  // Kill existing if any
  stopProcess(projectId);

  // Inherit full PATH so node/npm/bun/etc are available even when launched via LaunchAgent
  const env = {
    ...process.env,
    PATH: [
      '/Users/jon/.bun/bin',
      '/Users/jon/.npm-global/bin',
      '/opt/homebrew/bin',
      '/usr/local/bin',
      process.env.PATH,
    ].filter(Boolean).join(':'),
  };

  const proc = Bun.spawn(['sh', '-c', command], {
    cwd: path,
    stdout: 'pipe',
    stderr: 'pipe',
    stdin: 'ignore',
    env,
  });

  const managed: ManagedProcess = {
    proc,
    projectId,
    projectPath: path,
    command,
    buffer: [],
    listeners: new Set(),
    startedAt: Date.now(),
    autoRestart: options?.autoRestart ?? false,
    restartCount: 0,
    crashTimes: [],
    restarting: false,
  };

  processes.set(projectId, managed);

  // Pipe stdout
  if (proc.stdout) {
    const reader = proc.stdout.getReader();
    const decoder = new TextDecoder();
    (async () => {
      try {
        while (true) {
          const { done, value } = await reader.read();
          if (done) break;
          const text = decoder.decode(value, { stream: true });
          pushLines(managed, text);
        }
      } catch { /* process ended */ }
    })();
  }

  // Pipe stderr
  if (proc.stderr) {
    const reader = (proc.stderr as ReadableStream<Uint8Array>).getReader();
    const decoder = new TextDecoder();
    (async () => {
      try {
        while (true) {
          const { done, value } = await reader.read();
          if (done) break;
          const text = decoder.decode(value, { stream: true });
          pushLines(managed, text);
        }
      } catch { /* process ended */ }
    })();
  }

  // Handle exit
  proc.exited.then((exitCode) => {
    const current = processes.get(projectId);
    if (!current || current.proc !== proc) return; // stale reference

    if (exitCode !== 0 && !current.restarting) {
      // Unexpected crash
      pushEvent(current, 'crash', { projectId, exitCode, restartCount: current.restartCount });
      pushLines(current, `\n[Process crashed with exit code ${exitCode}]\n`);

      if (current.autoRestart) {
        handleCrash(current);
        return;
      }
    } else if (!current.restarting) {
      pushLines(current, '\n[Process exited]\n');
    }

    // Close SSE listeners if not restarting
    if (!current.restarting) {
      for (const ctrl of current.listeners) {
        try { ctrl.close(); } catch { /* ok */ }
      }
      current.listeners.clear();
    }
  });

  return true;
}

function handleCrash(managed: ManagedProcess) {
  const now = Date.now();
  managed.crashTimes.push(now);

  // Filter to crashes within the restart window
  managed.crashTimes = managed.crashTimes.filter(t => now - t < RESTART_WINDOW_MS);

  if (managed.crashTimes.length > MAX_RESTARTS) {
    pushEvent(managed, 'restart-failed', {
      projectId: managed.projectId,
      reason: `Too many restarts (${MAX_RESTARTS}) within ${RESTART_WINDOW_MS / 60000} minutes`,
    });
    pushLines(managed, `\n[Auto-restart disabled — ${MAX_RESTARTS} crashes in ${RESTART_WINDOW_MS / 60000} min]\n`);
    // Close listeners
    for (const ctrl of managed.listeners) {
      try { ctrl.close(); } catch { /* ok */ }
    }
    managed.listeners.clear();
    return;
  }

  managed.restartCount++;
  const delay = Math.min(1000 * Math.pow(2, managed.restartCount - 1), 30000);

  pushEvent(managed, 'restart', {
    projectId: managed.projectId,
    attempt: managed.restartCount,
    delayMs: delay,
  });
  pushLines(managed, `\n[Auto-restarting in ${delay}ms (attempt ${managed.restartCount})...]\n`);

  managed.restarting = true;

  setTimeout(() => {
    managed.restarting = false;
    const current = processes.get(managed.projectId);
    if (!current || current !== managed) return; // another start happened

    // Preserve listeners and state, spawn new process
    const cmdCheck = validateDevCommand(managed.command);
    if (!cmdCheck.valid) return;

    const restartEnv = {
      ...process.env,
      PATH: ['/Users/jon/.bun/bin', '/Users/jon/.npm-global/bin', '/opt/homebrew/bin', '/usr/local/bin', process.env.PATH].filter(Boolean).join(':'),
    };
    const proc = Bun.spawn(['sh', '-c', managed.command], {
      cwd: managed.projectPath,
      stdout: 'pipe',
      stderr: 'pipe',
      env: restartEnv,
      stdin: 'ignore',
    });

    managed.proc = proc;
    managed.startedAt = Date.now();

    // Re-pipe stdout
    if (proc.stdout) {
      const reader = proc.stdout.getReader();
      const decoder = new TextDecoder();
      (async () => {
        try {
          while (true) {
            const { done, value } = await reader.read();
            if (done) break;
            pushLines(managed, decoder.decode(value, { stream: true }));
          }
        } catch { /* process ended */ }
      })();
    }

    // Re-pipe stderr
    if (proc.stderr) {
      const reader = (proc.stderr as ReadableStream<Uint8Array>).getReader();
      const decoder = new TextDecoder();
      (async () => {
        try {
          while (true) {
            const { done, value } = await reader.read();
            if (done) break;
            pushLines(managed, decoder.decode(value, { stream: true }));
          }
        } catch { /* process ended */ }
      })();
    }

    // Re-attach exit handler
    proc.exited.then((exitCode) => {
      const cur = processes.get(managed.projectId);
      if (!cur || cur.proc !== proc) return;

      if (exitCode !== 0 && !cur.restarting) {
        pushEvent(cur, 'crash', { projectId: cur.projectId, exitCode, restartCount: cur.restartCount });
        pushLines(cur, `\n[Process crashed with exit code ${exitCode}]\n`);
        if (cur.autoRestart) {
          handleCrash(cur);
          return;
        }
      } else if (!cur.restarting) {
        pushLines(cur, '\n[Process exited]\n');
      }

      if (!cur.restarting) {
        for (const ctrl of cur.listeners) {
          try { ctrl.close(); } catch { /* ok */ }
        }
        cur.listeners.clear();
      }
    });
  }, delay);
}

function pushEvent(managed: ManagedProcess, event: string, data: Record<string, unknown>) {
  const sseData = `event: ${event}\ndata: ${JSON.stringify(data)}\n\n`;
  for (const ctrl of managed.listeners) {
    try {
      ctrl.enqueue(new TextEncoder().encode(sseData));
    } catch {
      managed.listeners.delete(ctrl);
    }
  }
}

function pushLines(managed: ManagedProcess, text: string) {
  const lines = text.split('\n');
  for (const line of lines) {
    if (line.length === 0 && lines.length > 1) continue;
    managed.buffer.push(line);
    if (managed.buffer.length > MAX_BUFFER) {
      managed.buffer.shift();
    }
    // Fan out to SSE listeners
    const sseData = `data: ${JSON.stringify(line)}\n\n`;
    for (const ctrl of managed.listeners) {
      try {
        ctrl.enqueue(new TextEncoder().encode(sseData));
      } catch {
        managed.listeners.delete(ctrl);
      }
    }
  }
}

export function stopProcess(projectId: string): boolean {
  const managed = processes.get(projectId);
  if (!managed) return false;
  managed.restarting = false; // prevent auto-restart on intentional stop
  managed.autoRestart = false;
  try { managed.proc.kill(); } catch { /* already dead */ }
  for (const ctrl of managed.listeners) {
    try { ctrl.close(); } catch { /* ok */ }
  }
  managed.listeners.clear();
  processes.delete(projectId);
  return true;
}

export function getBuffer(projectId: string): string[] {
  return processes.get(projectId)?.buffer || [];
}

export function getStatus(projectId: string): { running: boolean; pid: number | null; startedAt: number | null; autoRestart: boolean; restartCount: number } {
  const managed = processes.get(projectId);
  if (!managed) return { running: false, pid: null, startedAt: null, autoRestart: false, restartCount: 0 };
  return {
    running: !managed.proc.killed,
    pid: managed.proc.pid,
    startedAt: managed.startedAt,
    autoRestart: managed.autoRestart,
    restartCount: managed.restartCount,
  };
}

export function setAutoRestart(projectId: string, enabled: boolean): boolean {
  const managed = processes.get(projectId);
  if (!managed) return false;
  managed.autoRestart = enabled;
  return true;
}

export function getAllProcesses(): ProcessInfo[] {
  const result: ProcessInfo[] = [];
  for (const [, managed] of processes) {
    result.push({
      projectId: managed.projectId,
      running: !managed.proc.killed,
      pid: managed.proc.pid,
      startedAt: managed.startedAt,
      autoRestart: managed.autoRestart,
      restartCount: managed.restartCount,
    });
  }
  return result;
}

export function subscribe(projectId: string): ReadableStream<Uint8Array> {
  const managed = processes.get(projectId);

  return new ReadableStream<Uint8Array>({
    start(controller) {
      if (managed) {
        managed.listeners.add(controller);
      }
    },
    cancel() {
      if (managed) {
        managed.listeners.delete(this as unknown as ReadableStreamDefaultController);
      }
    },
  });
}

export function cleanup() {
  for (const [id] of processes) {
    stopProcess(id);
  }
}

File: Sources/DevDock/Server/ProcessManager.swift
import Foundation

final class ProcessManager {
    static let shared = ProcessManager()

    struct StatusSnapshot: Codable {
        let projectId: String
        let running: Bool
        let pid: Int?
        let startedAt: Int64
        let autoRestart: Bool
        let restartCount: Int
    }

    private struct ManagedProcess {
        var process: Process
        let projectId: String
        let projectPath: String
        let command: String
        var startedAt: Int64
        var autoRestart: Bool
        var restartCount: Int
        var crashTimes: [Date]
        var restarting: Bool
    }

    private let queue = DispatchQueue(label: "devdock.process-manager")
    private let maxBuffer = 500
    private let maxConcurrent = 20
    private let maxRestarts = 3
    private let restartWindow: TimeInterval = 5 * 60

    private var processes: [String: ManagedProcess] = [:]
    private var buffers: [String: [String]] = [:]
    private var statuses: [String: StatusSnapshot] = [:]

    func startProcess(projectId: String, path: String, command: String, autoRestart: Bool = false) -> Bool {
        guard Security.validateDevCommand(command).valid else { return false }

        return queue.sync {
            if processes.count >= maxConcurrent && processes[projectId] == nil {
                return false
            }

            if let existing = processes[projectId] {
                stop(process: existing.process, force: false)
                processes.removeValue(forKey: projectId)
            }

            guard let process = spawnProcess(projectId: projectId, path: path, command: command) else {
                return false
            }

            let now = nowMillis()
            let managed = ManagedProcess(
                process: process,
                projectId: projectId,
                projectPath: path,
                command: command,
                startedAt: now,
                autoRestart: autoRestart,
                restartCount: 0,
                crashTimes: [],
                restarting: false
            )
            processes[projectId] = managed
            statuses[projectId] = StatusSnapshot(
                projectId: projectId,
                running: true,
                pid: Int(process.processIdentifier),
                startedAt: now,
                autoRestart: autoRestart,
                restartCount: 0
            )
            buffers[projectId] = []
            Queries.recordActivity(projectId: projectId, signal: "server_start")
            wireIO(for: process, projectId: projectId)
            wireExit(for: process, projectId: projectId)
            return true
        }
    }

    func stopProcess(projectId: String) -> Bool {
        queue.sync {
            guard let managed = processes.removeValue(forKey: projectId) else {
                statuses[projectId] = StatusSnapshot(projectId: projectId, running: false, pid: nil, startedAt: statuses[projectId]?.startedAt ?? 0, autoRestart: false, restartCount: statuses[projectId]?.restartCount ?? 0)
                return false
            }

            stop(process: managed.process, force: false)
            appendLine(projectId: projectId, line: "[Process exited]")
            statuses[projectId] = StatusSnapshot(
                projectId: projectId,
                running: false,
                pid: nil,
                startedAt: managed.startedAt,
                autoRestart: false,
                restartCount: managed.restartCount
            )
            Queries.recordActivity(projectId: projectId, signal: "server_stop")
            SSEHub.shared.close(projectId: projectId)
            return true
        }
    }

    func getBuffer(projectId: String) -> [String] {
        queue.sync { buffers[projectId] ?? [] }
    }

    func getStatus(projectId: String) -> StatusSnapshot {
        queue.sync {
            statuses[projectId] ?? StatusSnapshot(projectId: projectId, running: false, pid: nil, startedAt: 0, autoRestart: false, restartCount: 0)
        }
    }

    func setAutoRestart(projectId: String, enabled: Bool) -> Bool {
        queue.sync {
            guard var managed = processes[projectId] else { return false }
            managed.autoRestart = enabled
            processes[projectId] = managed
            statuses[projectId] = StatusSnapshot(
                projectId: projectId,
                running: true,
                pid: Int(managed.process.processIdentifier),
                startedAt: managed.startedAt,
                autoRestart: enabled,
                restartCount: managed.restartCount
            )
            return true
        }
    }

    func getAllProcesses() -> [StatusSnapshot] {
        queue.sync {
            processes.values.map {
                StatusSnapshot(
                    projectId: $0.projectId,
                    running: true,
                    pid: Int($0.process.processIdentifier),
                    startedAt: $0.startedAt,
                    autoRestart: $0.autoRestart,
                    restartCount: $0.restartCount
                )
            }
            .sorted { $0.projectId < $1.projectId }
        }
    }

    func cleanup() {
        let active = queue.sync {
            let values = Array(processes.values)
            processes.removeAll()
            return values
        }
        for managed in active {
            stop(process: managed.process, force: true)
            SSEHub.shared.close(projectId: managed.projectId)
        }
        SSEHub.shared.closeAll()
    }

    private func spawnProcess(projectId: String, path: String, command: String) -> Process? {
        let process = Process()
        process.currentDirectoryURL = URL(fileURLWithPath: path)
        process.executableURL = URL(fileURLWithPath: "/bin/zsh")
        process.arguments = ["-lc", command]

        var environment = ProcessInfo.processInfo.environment
        let extraPath = [
            "/Users/jon/.bun/bin",
            "/Users/jon/.npm-global/bin",
            "/opt/homebrew/bin",
            "/usr/local/bin",
            environment["PATH"],
        ].compactMap { $0 }.joined(separator: ":")
        environment["PATH"] = extraPath
        process.environment = environment

        let stdout = Pipe()
        let stderr = Pipe()
        process.standardOutput = stdout
        process.standardError = stderr
        process.standardInput = nil

        do {
            try process.run()
            return process
        } catch {
            appendLine(projectId: projectId, line: "[Failed to start process: \(error.localizedDescription)]")
            return nil
        }
    }

    private func wireIO(for process: Process, projectId: String) {
        guard let stdout = process.standardOutput as? Pipe,
              let stderr = process.standardError as? Pipe else { return }

        stdout.fileHandleForReading.readabilityHandler = { [weak self] handle in
            let data = handle.availableData
            guard !data.isEmpty else {
                handle.readabilityHandler = nil
                return
            }
            self?.handleOutput(data: data, projectId: projectId)
        }

        stderr.fileHandleForReading.readabilityHandler = { [weak self] handle in
            let data = handle.availableData
            guard !data.isEmpty else {
                handle.readabilityHandler = nil
                return
            }
            self?.handleOutput(data: data, projectId: projectId)
        }
    }

    private func wireExit(for process: Process, projectId: String) {
        process.terminationHandler = { [weak self] proc in
            self?.handleTermination(projectId: projectId, process: proc)
        }
    }

    private func handleOutput(data: Data, projectId: String) {
        let text = String(data: data, encoding: .utf8) ?? String(decoding: data, as: UTF8.self)
        queue.async {
            let normalized = text.replacingOccurrences(of: "\r\n", with: "\n").replacingOccurrences(of: "\r", with: "\n")
            let parts = normalized.split(separator: "\n", omittingEmptySubsequences: false)
            for part in parts {
                self.appendLine(projectId: projectId, line: String(part))
            }
        }
    }

    private func handleTermination(projectId: String, process: Process) {
        queue.async {
            guard var managed = self.processes[projectId],
                  managed.process.processIdentifier == process.processIdentifier else {
                return
            }

            let exitCode = Int(process.terminationStatus)
            let crashed = exitCode != 0

            if crashed {
                self.appendLine(projectId: projectId, line: "[Process crashed with exit code \(exitCode)]")
            } else {
                self.appendLine(projectId: projectId, line: "[Process exited]")
            }

            if crashed && managed.autoRestart {
                let now = Date()
                managed.crashTimes.append(now)
                managed.crashTimes = managed.crashTimes.filter { now.timeIntervalSince($0) < self.restartWindow }

                if managed.crashTimes.count <= self.maxRestarts {
                    managed.restarting = true
                    managed.restartCount += 1
                    self.processes[projectId] = managed
                    self.statuses[projectId] = StatusSnapshot(
                        projectId: projectId,
                        running: false,
                        pid: nil,
                        startedAt: managed.startedAt,
                        autoRestart: true,
                        restartCount: managed.restartCount
                    )

                    let delay = min(pow(2.0, Double(managed.restartCount - 1)), 30.0)
                    self.appendLine(projectId: projectId, line: "[Auto-restarting in \(Int(delay * 1000))ms (attempt \(managed.restartCount))...]")
                    self.scheduleRestart(managed, delay: delay)
                    return
                }

                self.appendLine(projectId: projectId, line: "[Auto-restart disabled — too many crashes]")
            }

            self.processes.removeValue(forKey: projectId)
            self.statuses[projectId] = StatusSnapshot(
                projectId: projectId,
                running: false,
                pid: nil,
                startedAt: managed.startedAt,
                autoRestart: false,
                restartCount: managed.restartCount
            )
            Queries.recordActivity(projectId: projectId, signal: "server_stop", metadata: ["exitCode": exitCode])
            SSEHub.shared.close(projectId: projectId)
        }
    }

    private func scheduleRestart(_ managed: ManagedProcess, delay: TimeInterval) {
        queue.asyncAfter(deadline: .now() + delay) {
            guard var current = self.processes[managed.projectId],
                  current.process.processIdentifier == managed.process.processIdentifier,
                  current.restarting else {
                return
            }

            guard let process = self.spawnProcess(projectId: current.projectId, path: current.projectPath, command: current.command) else {
                self.processes.removeValue(forKey: current.projectId)
                self.statuses[current.projectId] = StatusSnapshot(projectId: current.projectId, running: false, pid: nil, startedAt: current.startedAt, autoRestart: false, restartCount: current.restartCount)
                SSEHub.shared.close(projectId: current.projectId)
                return
            }

            current.process = process
            current.startedAt = self.nowMillis()
            current.restarting = false
            self.processes[current.projectId] = current
            self.statuses[current.projectId] = StatusSnapshot(
                projectId: current.projectId,
                running: true,
                pid: Int(process.processIdentifier),
                startedAt: current.startedAt,
                autoRestart: current.autoRestart,
                restartCount: current.restartCount
            )
            self.wireIO(for: process, projectId: current.projectId)
            self.wireExit(for: process, projectId: current.projectId)
        }
    }

    private func appendLine(projectId: String, line: String) {
        let trimmedLine = line
        var buffer = buffers[projectId] ?? []
        buffer.append(trimmedLine)
        if buffer.count > maxBuffer {
            buffer.removeFirst(buffer.count - maxBuffer)
        }
        buffers[projectId] = buffer
        if !trimmedLine.isEmpty {
            SSEHub.shared.publish(projectId: projectId, message: trimmedLine)
        }
    }

    private func stop(process: Process, force: Bool) {
        guard process.isRunning else { return }
        process.terminate()
        if !force {
            return
        }

        DispatchQueue.global().asyncAfter(deadline: .now() + 1) {
            if process.isRunning {
                kill(process.processIdentifier, SIGKILL)
            }
        }
    }

    private func nowMillis() -> Int64 {
        Int64(Date().timeIntervalSince1970 * 1000)
    }
}
